### PR TITLE
Add geom to VariableSpec

### PR DESCRIPTION
**Pull request by pchakraborty**
## Types of change(s)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Trivial change (affects only documentation or cleanup)
- [ ] Refactor (no functional changes, no api changes)

## Checklist
- [ ] Tested this change with a run of GEOSgcm
- [x] Ran the Unit Tests (`make tests`)

## Description

Added `geom` as an optional variable in `VariableSpec`. The case where this `geom` exists and someone specifies a `geom` via `make_itemspec` is not allowed.

### CODE REVIEW

**Review by tclune — CHANGES_REQUESTED** [review]
Found some issues with scenarios where no geom is passed down.   Discussed with @pchakraborty offline.

File: generic3g/specs/VariableSpec.F90 [hunk 1]
```diff
@@ -199,11 +203,14 @@ function make_ItemSpec_new(this, geom, vertical_grid, registry, rc) result(item_
 
       integer :: status
       type(ActualPtVector) :: dependencies
+      type(ESMF_Geom), allocatable :: geom_local
+
+      call this%pick_geom_(geom, geom_local, _RC)
```
Comment (tclune) [inline generic3g/specs/VariableSpec.F90:208]: One subtle problem.   If `pick_geom` fails, then the allocatable function result (`item_spec`) is not allocated.  This creates errors upstream that cannot be trapped.

The procedure now needs a temporary assignment of `item_spec = NullSpec()` except ... that triggers a known problem with gfortran and reallocation of poly items that means you have to deallocate it again before each of the allocations in the SELECT CASE construct.  Ugh.

I'll approve but please open a ticket to fix this upon your return.

**Review by tclune — APPROVED** [review]

File: generic3g/specs/VariableSpec.F90 (hunk 1 above)
Comment (pchakraborty) [reply generic3g/specs/VariableSpec.F90:208]: Created the issue.

I have been struggling with these Fortran subtleties all of last week :-(. My understanding of Fortran is not what I thought it was.

Comment (tclune) [reply generic3g/specs/VariableSpec.F90:208]: > I have been struggling with these Fortran subtleties all of last week :-(. My understanding of Fortran is not what I thought it was.

It is not that your understanding was inadequate, but rather that the language (and my use of it) has opened new ways for things to fail.   Functions that can fail are really a bad idea.  But I like functional programming, and keep trying to apply it in a software ecosystem that is saturated with failure conditions.

**Pull request merged**